ssl/statem/extensions.c

```diff
@@ -1709,3 +1723,92 @@ static int final_psk(SSL *s, unsigned int context, int sent)
 
     return 1;
 }
+
+static int tls_init_compress_certificate(SSL *s, unsigned int context)
+{
+    s->ext.compress_certificate_rx = TLSEXT_comp_cert_none;
+    return 1;
+}
+
+/* The order these are put into the packet imply a preference order: [brotli, zlib, zstd] */
+static EXT_RETURN tls_construct_compress_certificate(SSL *s, WPACKET *pkt, unsigned int context,
+                                                     X509 *x, size_t chainidx)
+{
+#if !defined(OPENSSL_NO_COMP) && (defined(ZLIB) || defined(BROTLI) || defined(ZSTD))
+    if (s->options & SSL_OP_NO_CERTIFICATE_COMPRESSION)
+        return EXT_RETURN_NOT_SENT;
+
+    if (!WPACKET_put_bytes_u16(pkt, TLSEXT_TYPE_compress_certificate)
+            || !WPACKET_start_sub_packet_u16(pkt)
+            || !WPACKET_start_sub_packet_u8(pkt)
+# if defined(BROTLI)
+            || !WPACKET_put_bytes_u16(pkt, TLSEXT_comp_cert_brotli)
+# endif
+# if defined(ZLIB)
+            || !WPACKET_put_bytes_u16(pkt, TLSEXT_comp_cert_zlib)
+# endif
+# if defined(ZSTD)
+            || !WPACKET_put_bytes_u16(pkt, TLSEXT_comp_cert_zstd)
+# endif
+            || !WPACKET_close(pkt)
+            || !WPACKET_close(pkt)) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        return EXT_RETURN_FAIL;
+    }
+
+    s->ext.compress_certificate_tx = 1;
+    return EXT_RETURN_SENT;
+#else
+    return EXT_RETURN_NOT_SENT;
+#endif
+}
+
+int tls_parse_compress_certificate(SSL *s, PACKET *pkt, unsigned int context,
+                                   X509 *x, size_t chainidx)
+{
+#if !defined(OPENSSL_NO_COMP) && (defined(ZLIB) || defined(BROTLI) || defined(ZSTD))
+    PACKET supported_comp_algs;
+    unsigned int comp;
+
+    if (s->options & SSL_OP_NO_CERTIFICATE_COMPRESSION)
+        return 1;
+
+    if (s->ext.compress_certificate_rx != TLSEXT_comp_cert_none) {
+        SSLfatal(s, SSL_AD_DECODE_ERROR, SSL_R_DUPLICATE_EXTENSION);
```

Comment:
`tls_collect_extensions` checks for duplicates so this is a "should not happen" scenario. Perhaps this should be an `ossl_assert`?